File: src/wasm.js
const { promises: { readFile } } = require('fs');

async function main() {
    const wasm = await readFile('color-control.wasm');
    let heap;
    function toString(offset) {
        let str = "";
        while (heap[offset]) {
            str += String.fromCharCode(heap[offset++]);
        }

    }
    let first = true;
    const { instance: { exports } } = await WebAssembly.instantiate(wasm, {
        env: {
            puts(offset) {
                console.log(toString(offset));
            },
            //    ◢■■◣
            //  ◢■◤  ◥■◣
            // ■■      ■■
            //  ◥■◣  ◢■◤
            //    ◥■■◤
            on_update(ptr) {
                const pixels = heap.subarray(ptr);
                let str = "";
                if (first) {
                    first = false;
                }
                else {
                    str = "\x1b[5A";
                }
                str +=
                    `   \x1b[38;2;${pixels[0]};${pixels[1]};${pixels[2]}m◢■■◣\x1b[0m\n` +
                    ` \x1b[38;2;${pixels[21]};${pixels[22]};${pixels[23]}m◢■◤\x1b[0m  \x1b[38;2;${pixels[3]};${pixels[4]};${pixels[5]}m◥■◣\x1b[0m\n` +
                    `\x1b[38;2;${pixels[18]};${pixels[19]};${pixels[20]}m■■\x1b[0m      \x1b[38;2;${pixels[6]};${pixels[7]};${pixels[8]}m■■\x1b[0m\n` +
                    ` \x1b[38;2;${pixels[15]};${pixels[16]};${pixels[17]}m◥■◣\x1b[0m  \x1b[38;2;${pixels[9]};${pixels[10]};${pixels[11]}m◢■◤\x1b[0m\n` +
                    `   \x1b[38;2;${pixels[12]};${pixels[13]};${pixels[14]}m◥■■◤\x1b[0m`
                console.log(str);
            },
            on_pin(bank, pin, state) {
                console.log("on_pin", { bank, pin, state });
            },
            on_error(code, msg) {
                const err = new Error(toString(msg));
                err.code = code;
                throw err;
            },
            on_delay(ms) {
                const end = Date.now() + ms;
                while (Date.now() < end);
            }
        }
    });
    heap = new Uint8Array(exports.memory.buffer);

    const code = await readFile("power-rainbow.code");
    new Uint8Array(exports.memory.buffer, exports.get_code(), 1024).set(code.subarray(4, code.length - 2));
    const result = exports.color_control(code.length);
    console.log({ result });
}
main();

File: src/wasm.c
#include <emscripten.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include "./libs/color-control.c"

extern void on_update(uint8_t* pixels);
extern void on_pin(uint8_t bank, uint8_t pin, bool state);
extern void on_error(uint32_t code, const char* msg);
extern void on_delay(uint32_t ms);

static uint8_t code[1024];
static vm_t vm = (vm_t){.on_update = on_update,
                        .on_pin = on_pin,
                        .on_error = on_error,
                        .on_delay = on_delay,
                        .vars = {0, 0, 0, 0, 0, 0, 0, 0},
                        .pc = code};

EMSCRIPTEN_KEEPALIVE
uint8_t* get_code() {
  return code;
}

EMSCRIPTEN_KEEPALIVE
uint32_t color_control(uint16_t len) {
  return eval(&vm);
}
